File: Calculator.cpp
#include <iostream>
#include <cmath>
#include <limits>

using namespace std;

void displayMenu() {
    cout << "\nAdvanced Calculator" << endl;
    cout << "1. Addition" << endl;
    cout << "2. Subtraction" << endl;
    cout << "3. Multiplication" << endl;
    cout << "4. Division" << endl;
    cout << "5. Power" << endl;
    cout << "6. Square Root" << endl;
    cout << "7. Factorial" << endl;
    cout << "8. Logarithm (Base 10)" << endl;
    cout << "9. Trigonometric Functions" << endl;
    cout << "10. Exit" << endl;
}

double getInput(const string &prompt) {
    double value;
    cout << prompt;
    while (!(cin >> value)) {
        cout << "Error: Invalid input! Please enter a numeric value: ";
        cin.clear();
        cin.ignore(numeric_limits<streamsize>::max(), '\n');
    }
    return value;
}

int main() {
    while (true) {
        displayMenu();

        int choice;
        cout << "\nEnter your choice (1-10): ";
        cin >> choice;

        if (cin.fail() || choice < 1 || choice > 10) {
            cout << "Error: Invalid choice! Please enter a number between 1 and 10." << endl;
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            continue;
        }

        if (choice == 10) {
            cout << "Exiting calculator. Goodbye!" << endl;
            break;
        }

        switch (choice) {
            case 1: { // Addition
                double a = getInput("Enter the first number: ");
                double b = getInput("Enter the second number: ");
                cout << "Result: " << (a + b) << endl;
                break;
            }
            case 2: { // Subtraction
                double a = getInput("Enter the first number: ");
                double b = getInput("Enter the second number: ");
                cout << "Result: " << (a - b) << endl;
                break;
            }
            case 3: { // Multiplication
                double a = getInput("Enter the first number: ");
                double b = getInput("Enter the second number: ");
                cout << "Result: " << (a * b) << endl;
                break;
            }
            case 4: { // Division
                double a = getInput("Enter the numerator: ");
                double b = getInput("Enter the denominator: ");
                if (b == 0) {
                    cout << "Error: Division by zero!" << endl;
                } else {
                    cout << "Result: " << (a / b) << endl;
                }
                break;
            }
            case 5: { // Power
                double base = getInput("Enter the base: ");
                double exponent = getInput("Enter the exponent: ");
                cout << "Result: " << pow(base, exponent) << endl;
                break;
            }
            case 6: { // Square Root
                double a = getInput("Enter the number: ");
                if (a < 0) {
                    cout << "Error: Square root of a negative number is not real!" << endl;
                } else {
                    cout << "Result: " << sqrt(a) << endl;
                }
                break;
            }
            case 7: { // Factorial
                int n = static_cast<int>(getInput("Enter a non-negative integer: "));
                if (n < 0) {
                    cout << "Error: Factorial of a negative number is undefined!" << endl;
                } else {
                    long long factorial = 1;
                    for (int i = 1; i <= n; ++i) {
                        factorial *= i;
                    }
                    cout << "Result: " << factorial << endl;
                }
                break;
            }
            case 8: { // Logarithm
                double a = getInput("Enter the number: ");
                if (a <= 0) {
                    cout << "Error: Logarithm is undefined for zero or negative numbers!" << endl;
                } else {
                    cout << "Result: " << log10(a) << endl;
                }
                break;
            }
            case 9: { // Trigonometric Functions
                cout << "\nTrigonometric Functions:" << endl;
                cout << "1. Sine" << endl;
                cout << "2. Cosine" << endl;
                cout << "3. Tangent" << endl;
                int trigChoice;
                cout << "Enter your choice (1-3): ";
                cin >> trigChoice;

                if (cin.fail() || trigChoice < 1 || trigChoice > 3) {
                    cout << "Error: Invalid choice!" << endl;
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    break;
                }

                double angle = getInput("Enter the angle in degrees: ");
                double radians = angle * M_PI / 180.0;

                if (trigChoice == 1) {
                    cout << "Result: " << sin(radians) << endl;
                } else if (trigChoice == 2) {
                    cout << "Result: " << cos(radians) << endl;
                } else if (trigChoice == 3) {
                    cout << "Result: " << tan(radians) << endl;
                }
                break;
            }
            default:
                cout << "Error: Unknown error occurred!" << endl;
        }
    }

    return 0;
}

File: Calculator.py
import math

def calculator():
    while True:
        print("\nAdvanced Calculator")
        print("1. Addition")
        print("2. Subtraction")
        print("3. Multiplication")
        print("4. Division")
        print("5. Power")
        print("6. Square Root")
        print("7. Factorial")
        print("8. Logarithm (Base 10)")
        print("9. Trigonometric Functions")
        print("10. Exit")
        
        try:
            choice = int(input("\nEnter your choice (1-10): "))
            
            if choice == 1:
                a, b = map(float, input("Enter two numbers separated by a space: ").split())
                print(f"Result: {a + b}")
            
            elif choice == 2:
                a, b = map(float, input("Enter two numbers separated by a space: ").split())
                print(f"Result: {a - b}")
            
            elif choice == 3:
                a, b = map(float, input("Enter two numbers separated by a space: ").split())
                print(f"Result: {a * b}")
            
            elif choice == 4:
                a, b = map(float, input("Enter two numbers separated by a space: ").split())
                if b == 0:
                    print("Error: Division by zero!")
                else:
                    print(f"Result: {a / b}")
            
            elif choice == 5:
                a, b = map(float, input("Enter base and exponent separated by a space: ").split())
                print(f"Result: {math.pow(a, b)}")
            
            elif choice == 6:
                a = float(input("Enter the number: "))
                if a < 0:
                    print("Error: Square root of negative number is not real!")
                else:
                    print(f"Result: {math.sqrt(a)}")
            
            elif choice == 7:
                a = int(input("Enter a non-negative integer: "))
                if a < 0:
                    print("Error: Factorial of a negative number is undefined!")
                else:
                    print(f"Result: {math.factorial(a)}")
            
            elif choice == 8:
                a = float(input("Enter the number: "))
                if a <= 0:
                    print("Error: Logarithm is undefined for zero or negative numbers!")
                else:
                    print(f"Result: {math.log10(a)}")
            
            elif choice == 9:
                print("Trigonometric Functions:")
                print("1. Sine")
                print("2. Cosine")
                print("3. Tangent")
                trig_choice = int(input("Enter your choice (1-3): "))
                angle = float(input("Enter angle in degrees: "))
                radians = math.radians(angle)
                
                if trig_choice == 1:
                    print(f"Result: {math.sin(radians)}")
                elif trig_choice == 2:
                    print(f"Result: {math.cos(radians)}")
                elif trig_choice == 3:
                    print(f"Result: {math.tan(radians)}")
                else:
                    print("Invalid choice!")
            
            elif choice == 10:
                print("Exiting calculator. Goodbye!")
                break
            
            else:
                print("Invalid choice! Please enter a number between 1 and 10.")
        
        except ValueError:
            print("Error: Invalid input! Please enter numeric values.")

if __name__ == "__main__":
    calculator()
